torch/_inductor/codegen/cpp_micro_gemm.py

```diff
@@ -523,6 +527,157 @@ def codegen_define(self, kernel: CppTemplateKernel) -> str:
+    {{kernel.assert_function}}(K % {{block_k}} == 0, "K dimension must be multiple of {{block_k}}");
+    // TODO(jgong5): loop unroll for M and N
+    for (int64_t m = 0; m < M; m += {{block_m}}) {
+        int64_t block_m = std::min<int64_t>(M - m, {{block_m}});
+        for (int64_t n = 0; n < N; n += {{block_n}}) {
+            if (block_m == {{block_m}}) {
+                {{kernel_name}}_kernel<{{block_m}}, {{block_n}}, accum>(
+                    A + m * lda,
+                    B + n,
+                    S + n,
+                    C + m * ldc + n,
+                    K,
+                    lda,
+                    ldb,
+                    ldc
+                );
+            } else {
+                switch (block_m) {
+{%- for b in range(block_m - 1, 0, -1) %}
+                case {{b}}:
+                    {{kernel_name}}_kernel<{{b}}, {{block_n}}, accum>(
+                        A + m * lda,
+                        B + n,
+                        S + n,
+                        C + m * ldc + n,
+                        K,
+                        lda,
+                        ldb,
+                        ldc
+                    );
+                    break;
+{%- endfor %}
+                default:
+                    {{kernel.assert_function}}(false, "Unsupported block_m: {{block_m}}");
+                }
+            }
+        }
+    }
+}
+"""
+
+    TEMPLATE_KERNEL = r"""
+template <int64_t BLOCK_M, int64_t BLOCK_N, bool accum>
+inline void {{kernel_name}}_kernel(
+    const {{input_t}}* {{restrict_keyword}} A,
+    const {{input2_t}}* {{restrict_keyword}} B,
+    const {{input_t}}*{{restrict_keyword}} S,
+    {{output_t}}* {{restrict_keyword}} C,
+    int64_t K,
+    int64_t lda,
+    int64_t ldb,
+    int64_t ldc
+) {
+    constexpr auto VLEN = 32;
+    constexpr auto ROWS = BLOCK_M;
+    constexpr auto COLS = BLOCK_N / VLEN;
+
+    __m512h va;
+    __m512h vb[COLS];
+    __m512h scales[COLS];
+    __m512h vc[ROWS * COLS];
+
+    auto load_scales = [&](int i) {
+        scales[i] = _mm512_loadu_ph((__m512h*)(S + i * VLEN));
+    };
+    c10::ForcedUnroll<COLS>{}(load_scales);
+
+    auto loadc = [&](auto i) {
+        vc[i] = _mm512_set1_ph(0.0f);
+    };
+    c10::ForcedUnroll<ROWS * COLS>{}(loadc);
+
+    auto compute = [&, COLS](auto i, int k) {
+        constexpr int row = i / COLS;
+        constexpr int col = i % COLS;
+
+        if constexpr (col == 0) {
+            va = _mm512_set1_ph(*((_Float16*)(A + row * lda + k)));
+        }
+
+        if constexpr (row == 0) {
+            // Convert VLEN int8 elements to int16, then fp16, and then apply scale
+            auto int8_vector = _mm256_loadu_si256((__m256i*)(B + k * ldb + col * VLEN));
+            vb[col] =  _mm512_cvtepi16_ph(_mm512_cvtepi8_epi16(int8_vector));
+            vb[col] = _mm512_mul_ph(vb[col], scales[col]);
+        }
+
+        vc[i] = _mm512_fmadd_ph(va, vb[col], vc[i]);
```

Comment:
Do we need GCC version check to use this intrinsic?

Reply:
In general, we need a mechanism to query compiler support before even deciding to use a kernel that uses some specific ISA. 

Could we check it in `cpp_builder.py` by searching for the output of `gcc -Q --help=target | grep avx512fp16` (with gcc), for example? Thanks!

Reply:
I think there has the mechanism, for example: https://github.com/pytorch/pytorch/blob/cf6d1e68244c8376e72e536dba0acae913a038bd/torch/_inductor/cpu_vec_isa.py#L242